File: packages/temper-placer/src/temper_placer/pipeline/derivation.py
"""
Physics-based constraint derivation for PCB placement.

This module derives geometric placement constraints from high-level
physical performance specifications (EMI, Thermal, Signal Integrity).

The derivation compute (EMI `sqrt(area) * 0.8`, thermal `power * 2.0`,
signal-integrity `max_len / 1.5`, the mains-voltage classification
boundaries and the `_min_clearance` key extraction) delegates to the
``temper-orchestration`` crate (``temper_orchestration.derive_emi_max_dist``
/ ``derive_thermal_clearance`` / ``derive_si_max_placement_dist`` /
``mains_voltage_to_class_code`` / ``extract_min_clearance``), pinned
bit-identically by ``tests/pipeline/test_pipeline_feasibility_rust_differential.py``
(oracle: ``tests/pipeline/_pipeline_feasibility_py_oracle.py``).
"""

from __future__ import annotations

import warnings
from typing import TYPE_CHECKING, Any

import temper_orchestration as _rs

if TYPE_CHECKING:
    from temper_placer.core.netlist import Netlist
    from temper_placer.core.specification import PcbSpecification

from temper_placer.core.net_types import VoltageClass

# IEC 60335-1 voltage classes, indexed by `mains_voltage_to_class_code`
# (0=LOW_VOLTAGE, 1=MAINS_120V, 2=MAINS_240V, 3=HIGH_VOLTAGE).
_VOLTAGE_CLASS_BY_CODE = {
    0: VoltageClass.LOW_VOLTAGE,
    1: VoltageClass.MAINS_120V,
    2: VoltageClass.MAINS_240V,
    3: VoltageClass.HIGH_VOLTAGE,
}


def _mains_voltage_to_class(voltage_v: float) -> VoltageClass:
    """Map a mains voltage in volts to the nearest IEC 60335-1 voltage class.

    The boundary comparisons (`<= 50` / `<= 130` / `<= 264`, NaN falling
    through to HIGH_VOLTAGE) are the Rust kernel `mains_voltage_to_class_code`;
    the code-to-pyclass mapping stays here.
    """
    return _VOLTAGE_CLASS_BY_CODE[_rs.mains_voltage_to_class_code(voltage_v)]


def derive_constraints_from_spec(
    spec: PcbSpecification,
    netlist: Netlist,  # noqa: ARG001
) -> dict[str, Any]:
    """
    Derive geometric constraints from physical specifications.

    Returns a dictionary of derived parameters (e.g. max distances).
    """
    derived = {}

    # 1. EMI -> Max Distance and Max Area
    for loop_name, max_area in spec.emi.max_loop_area_mm2.items():
        # L = sqrt(Area). Max side length of a square loop. Conservative
        # estimate for max component spacing (center-to-center), assuming
        # 20% routing overhead. The `sqrt(area) * 0.8` expression is the
        # Rust kernel `derive_emi_max_dist` (IEEE sqrt = CPython math.sqrt,
        # then the same `* 0.8`).
        derived[f"{loop_name}_max_dist"] = _rs.derive_emi_max_dist(float(max_area))
        # Store the max area directly for quality metric scoring
        derived[f"{loop_name}_max_area_mm2"] = max_area

    # 2. Thermal -> Min Spacing
    # Simple model: heat sources should be spaced to avoid thermal overlap
    # Required spacing proportional to power dissipation
    power_map = spec.thermal.power_dissipation
    for ref, power in power_map.items():
        # Heuristic: 2mm per Watt spacing
        # The `power * 2.0` product is the Rust kernel
        # `derive_thermal_clearance`.
        derived[f"{ref}_min_clearance"] = _rs.derive_thermal_clearance(float(power))

    # 3. Signal Integrity -> Max Length
    for net_name, max_len in spec.signal_integrity.max_length_mm.items():
        # Max placement distance should be less than max length
        # Assuming 1.5x routing overhead (Manhattan + detours)
        # The `max_len / 1.5` quotient is the Rust kernel
        # `derive_si_max_placement_dist`.
        derived[f"{net_name}_max_placement_dist"] = _rs.derive_si_max_placement_dist(
            float(max_len)
        )

    # 4. Safety -> Isolation (Creepage/Clearance)
    if spec.safety is not None:
        vc = _mains_voltage_to_class(spec.safety.mains_voltage_v)
        derived["hv_lv_isolation_mm"] = vc.get_clearance_mm(
            pollution_degree=spec.safety.pollution_degree
        )
    else:
        # Default to 6.5mm for reinforced isolation (340V)
        warnings.warn(
            "No safety spec provided — falling back to hardcoded 6.5mm isolation.",
            stacklevel=2,
        )
        derived["hv_lv_isolation_mm"] = 6.5

    return derived


def apply_derived_constraints(
    netlist: Netlist,
    derived: dict[str, Any],
    pcl_constraints: Any = None,
) -> Any:
    """
    Apply derived constraints back to PCL constraint collection.

    When pcl_constraints is provided, synthesized constraints from
    derivation are added to it. Returns the modified collection or
    netlist fallback.

    This resolves the TODO at derivation.py:65 — back-propagation
    of derived parameters to the PCL constraint IR.
    """
    if pcl_constraints is None:
        return netlist

    from temper_placer.pcl.constraints import ConstraintTier, SeparatedConstraint

    for key, value in derived.items():
        # The `_min_clearance` suffix test and the ref extraction
        # (Python `str.replace` all-occurrence semantics) are the Rust
        # kernel `extract_min_clearance`; the PCL object construction is
        # Python IR marshalling.
        extracted = _rs.extract_min_clearance(key, float(value))
        if extracted is not None:
            ref, min_distance = extracted
            pcl_constraints.add(
                SeparatedConstraint(
                    a=ref,
                    b="*",
                    min_distance_mm=float(min_distance),
                    tier=ConstraintTier.STRONG,
                    because=f"Derived from thermal spec: {ref} min clearance {value}mm",
                )
            )

    return pcl_constraints

File: packages/temper-quality-oracle/src/derivation.rs
/// Constraint derivation from PCB specification.
///
/// Ports `derive_constraints_from_spec()` from Python:
/// - EMI → max component spacing (sqrt(area) * 0.8)
/// - Thermal → min clearance (power * 2.0)
/// - Signal Integrity → max placement distance (max_len / 1.5)
/// - Safety → HV-LV isolation (6.5mm default)

use crate::types::{DerivedConstraints, NetClassification, PcbSpecification};

pub fn derive(spec: &PcbSpecification, _classifications: &[NetClassification]) -> DerivedConstraints {
    let mut constraints = DerivedConstraints {
        hv_lv_isolation_mm: 6.5,
        ..Default::default()
    };

    for (loop_name, max_area) in &spec.max_loop_area_mm2 {
        let max_side = max_area.sqrt();
        constraints.loop_spacing.insert(
            format!("{}_max_dist", loop_name),
            max_side * 0.8,
        );
    }

    for (ref_des, power) in &spec.power_dissipation {
        constraints.thermal_clearances.insert(
            format!("{}_min_clearance", ref_des),
            power * 2.0,
        );
    }

    for (net_name, max_len) in &spec.max_length_mm {
        constraints.si_max_placement_dist.insert(
            format!("{}_max_placement_dist", net_name),
            max_len / 1.5,
        );
    }

    constraints
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    #[test]
    fn test_derive_emi_loop_spacing() {
        let spec = PcbSpecification {
            name: "test".into(),
            max_loop_area_mm2: HashMap::from([
                ("commutation_loop".into(), 80.0),
            ]),
            power_dissipation: HashMap::new(),
            max_length_mm: HashMap::new(),
            max_junction_temp_c: 125.0,
            ambient_temp_c: 40.0,
        };
        let classifications = vec![];
        let derived = derive(&spec, &classifications);
        let expected = 80.0_f64.sqrt() * 0.8;
        assert!((derived.loop_spacing["commutation_loop_max_dist"] - expected).abs() < 1e-6);
    }

    #[test]
    fn test_derive_thermal_clearance() {
        let spec = PcbSpecification {
            name: "test".into(),
            max_loop_area_mm2: HashMap::new(),
            power_dissipation: HashMap::from([
                ("Q1".into(), 15.0),
                ("Q2".into(), 10.0),
            ]),
            max_length_mm: HashMap::new(),
            max_junction_temp_c: 125.0,
            ambient_temp_c: 40.0,
        };
        let derived = derive(&spec, &[]);
        assert!((derived.thermal_clearances["Q1_min_clearance"] - 30.0).abs() < 1e-6);
        assert!((derived.thermal_clearances["Q2_min_clearance"] - 20.0).abs() < 1e-6);
    }

    #[test]
    fn test_derive_si_max_placement_dist() {
        let spec = PcbSpecification {
            name: "test".into(),
            max_loop_area_mm2: HashMap::new(),
            power_dissipation: HashMap::new(),
            max_length_mm: HashMap::from([
                ("CLK".into(), 100.0),
            ]),
            max_junction_temp_c: 125.0,
            ambient_temp_c: 40.0,
        };
        let derived = derive(&spec, &[]);
        let expected = 100.0 / 1.5;
        assert!((derived.si_max_placement_dist["CLK_max_placement_dist"] - expected).abs() < 1e-6);
    }

    #[test]
    fn test_derive_hv_lv_isolation_default() {
        let spec = PcbSpecification {
            name: "test".into(),
            max_loop_area_mm2: HashMap::new(),
            power_dissipation: HashMap::new(),
            max_length_mm: HashMap::new(),
            max_junction_temp_c: 125.0,
            ambient_temp_c: 40.0,
        };
        let derived = derive(&spec, &[]);
        assert!((derived.hv_lv_isolation_mm - 6.5).abs() < 1e-6);
    }

    #[test]
    fn test_derive_empty_spec() {
        let spec = PcbSpecification {
            name: "empty".into(),
            max_loop_area_mm2: HashMap::new(),
            power_dissipation: HashMap::new(),
            max_length_mm: HashMap::new(),
            max_junction_temp_c: 125.0,
            ambient_temp_c: 40.0,
        };
        let derived = derive(&spec, &[]);
        assert!(derived.loop_spacing.is_empty());
        assert!(derived.thermal_clearances.is_empty());
        assert!(derived.si_max_placement_dist.is_empty());
    }

    #[test]
    fn test_derived_distances_non_negative() {
        let spec = PcbSpecification {
            name: "test".into(),
            max_loop_area_mm2: HashMap::from([
                ("tiny_loop".into(), 0.01),
            ]),
            power_dissipation: HashMap::new(),
            max_length_mm: HashMap::from([
                ("short".into(), 0.1),
            ]),
            max_junction_temp_c: 125.0,
            ambient_temp_c: 40.0,
        };
        let derived = derive(&spec, &[]);
        for (_, v) in &derived.loop_spacing {
            assert!(*v >= 0.0, "loop spacing must be non-negative");
        }
        for (_, v) in &derived.si_max_placement_dist {
            assert!(*v >= 0.0, "SI placement dist must be non-negative");
        }
    }
}
